src/solvers/petsc_linear_solver.C

```diff
@@ -356,28 +356,36 @@ PetscLinearSolver<T>::solve (SparseMatrix<T> &  matrix_in,
                              SparseMatrix<T> &  precond_in,
                              NumericVector<T> & solution_in,
                              NumericVector<T> & rhs_in,
-                             const double tol,
-                             const unsigned int m_its)
+                             const std::optional<double> tol,
+                             const std::optional<unsigned int> m_its)
 {
   LOG_SCOPE("solve()", "PetscLinearSolver");
 
+  const double rel_tol = this->get_real_solver_setting("rel_tol", tol);
+  const double abs_tol = this->get_real_solver_setting("abs_tol", std::nullopt, (double)PETSC_DEFAULT);
```

Comment:
I'd prefer `static_cast<Real>(PETSC_DEFAULT)` assuming that works instead of the C-style cast with `double`.

Reply:
I'm cool with either `static_cast` or with a construction rather than a cast ... but we should really figure out whether we need this all to be `double` or `Real`.  I thought PETSc still uses `double` for tolerances even when the underlying `PetscReal` has been switched to `float` or `__float128`, so we'd want `double` here ... but if I'm wrong then I'd really worry about whether we can convert `PETSC_DEFAULT` from one of these to another without inadvertently trashing it.

Reply:
[KSPSetTolerances](https://petsc.org/release/manualpages/KSP/KSPSetTolerances/) takes `PetscReal`. My assumption was that we make our `Real` match `PetscReal`, but I don't know exactly how you handled this.

Reply:
Okay, I was wrong: it's `PetscReal` for everything in https://petsc.org/release/manualpages/KSP/KSPSetTolerances/

So we want to be careful to use `Real` for anything that we can't be certain will be happy about a conversion, like `PETSC_DEFAULT`.

Reply:
For loose definitions of "match"; IIRC they use pure `__float128` in the quad precision case, while we use `boost::multiprecision::float128` to get some `std::` function overloads from there.

Reply:
Anyway, once the `PETSC_DEFAULT` stuff is changed from `double` to `Real`, I'm happy with this.

Reply:
Ahh okay, github does not load new conversations autoamtically. I'll make it Real.